Check that value is preserved in v_head

v0d1ch opened the pull request:
## Why

According to spec `5.5. rule 6 -> value is preserved` there should be a validator check to make sure the head value is preserved. Note that `checkAbort` and `checkFanout` head transitions should not contain this check.

To check before merging:
* [x] CHANGELOG is up to date
* [x] Up to date with master

v0d1ch committed:
Add failing mutation and start a validator fix

v0d1ch committed:
Shuffle things around and add value preservation test for contest tx

v0d1ch committed:
Update changelog

ffakenz commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -98,9 +98,12 @@ validator (_party, _headScriptHash, _commit, headId) r ctx =
-      traceIfFalse "ST is missing in the output" (hasST headId outputs)
+      traceIfFalse "ST is missing in the output" (hasST headId headOutputValue)
  where
-  outputs = foldMap txOutValue $ txInfoOutputs $ scriptContextTxInfo ctx
```
interesting

ffakenz commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -193,7 +193,8 @@ checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPer
 
   (expectedChangeValue, collectedCommits, nTotalCommits) =
-    traverseInputs
+    foldr
```
nice one!

ch1bo commented on hydra-plutus/src/Hydra/Contract/Util.hs:
```diff
@@ -46,3 +47,14 @@ mustNotMintOrBurn TxInfo{txInfoMint} =
+mustPreserveValue outValue headOutValue =
+  traceIfFalse "head value is not preserved" $
+    outValue == headOutValue
+{-# INLINEABLE mustPreserveValue #-}
```
Not sure if having this in a function is any helpful if you need to provide also the two values to compare.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Util.hs:
```diff
@@ -46,3 +47,14 @@ mustNotMintOrBurn TxInfo{txInfoMint} =
+
+headOutputValue :: [TxOut] -> Value
+headOutputValue (headOutput : _outputs) = txOutValue headOutput
+headOutputValue _ = traceError "does not have at least head output"
```
This function itself has nothing to do with a "head" output. It's `getFirstOutputValue`. But I also don't see much value in having a function for this.

v0d1ch committed:
Remove addChangeOutput

v0d1ch committed:
Remove headOutputValue

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -357,7 +359,9 @@ checkContest ctx contestationDeadline parties closedSnapshotNumber sig headId =
  where
+  headOutValue = txOutValue . head $ txInfoOutputs txInfo
   outValue =
     maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
```
```suggestion
    && mustPreserveValue
 where 
  mustPreserveValue =
    traceIfFalse "head value is not preserved" $
      val == val'
  
  val' = txOutValue . head $ txInfoOutputs txInfo
  
  val = maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
```

Could: follow a bit closer the terminology of the paper. Also, functions don't really help here when they are getting inlined anways (IMO).

ch1bo reviewed (commented):
Changes look good, but what happened to the `close` and `contest` benchmarks? Only 10 parties there is a bit odd..

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -398,7 +398,7 @@ forAllFanout action =
             & label ("Fanout size: " <> prettyLength (countAssets $ txOuts' tx))
  where
-  maxSupported = 30
+  maxSupported = 39
```
Benchmarks show `45`, but there might be some delta between the generators used in `tx-cost` and these tests here. It's fine to only bump it to 39, although we could try higher numbers to test the limits more often.

v0d1ch committed:
Address PR review comments

v0d1ch committed:
Bump maxSupported outputs to 39

ch1bo committed:
Update tx-cost to see limits of close/contest

The transactions seem to be memory bound now.

ch1bo committed:
Identify what makes close/contest expensive: Value equality

The Eq Value instance is doing a lot of work to make two values equal
even if their actual contents (associative lists) are shuffled but consistent.

ch1bo reviewed (approved):
Now as we know why the close/contest is much smaller I can approve.

v0d1ch committed:
Reduce the supported outputs by one for the fanout tx
